patches/server/0066-Heavily-optimize-recipe-lookups.patch

```diff
@@ -0,0 +1,122 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Mykyta Komarn <nkomarn@hotmail.com>
+Date: Wed, 30 Sep 2020 18:17:51 -0700
+Subject: [PATCH] Heavily optimize recipe lookups
+
+Recipe lookups are now cached in CraftingManager. I also rewrote the stream replacement code to not create an ArrayList, rather just use the Collection that was already there. Additionally, an EMPTY_MAP variable was added to prevent bottlenecks during map creation, since that map is only ever iterated.
+
+NonNullList now uses GlueList internally as its delegate list, which should bring general improvements to all tile entities with an inventory.
+
+These changes knock off ~40ms of tick duration with a sample of ~7,700 running furnaces on a server.
+
+diff --git a/src/main/java/net/minecraft/server/CraftingManager.java b/src/main/java/net/minecraft/server/CraftingManager.java
+index de7d5ed8d6260ce5ee4164df29cdaf69f561045c..22b8166ebb8e407d49bb73cf09163a11d31d1666 100644
+--- a/src/main/java/net/minecraft/server/CraftingManager.java
++++ b/src/main/java/net/minecraft/server/CraftingManager.java
+@@ -25,6 +25,10 @@ public class CraftingManager extends ResourceDataJson {
+     private static final Logger LOGGER = LogManager.getLogger();
+     public Map<Recipes<?>, Object2ObjectLinkedOpenHashMap<MinecraftKey, IRecipe<?>>> recipes = ImmutableMap.of(); // CraftBukkit
+     private boolean d;
++    // Yatopia start
++    public static final Object2ObjectLinkedOpenHashMap<MinecraftKey, IRecipe<?>> EMPTY_MAP = new Object2ObjectLinkedOpenHashMap<>();
++    public static final Map<Recipes<?>, List> CACHE = new Object2ObjectLinkedOpenHashMap<>();
++    // Yatopia end
+ 
+     public CraftingManager() {
+         super(CraftingManager.a, "recipes");
+@@ -79,8 +83,8 @@ public class CraftingManager extends ResourceDataJson {
+     public <C extends IInventory, T extends IRecipe<C>> Optional<T> craft(Recipes<T> recipes, C c0, World world) {
+         // CraftBukkit start
+         // Yatopia start - replace stream
+-        List<IRecipe<C>> collection = new java.util.ArrayList<>(this.b(recipes).values());
+-        Optional<T> recipe = collection.isEmpty() ? Optional.empty() : recipes.a(collection.get(0), world, c0);
```

Comment:
Leave this as is. It doesnt do any benefit from the current code. Useless diff

Reply:
This is actually the part that brought the most performance impact. The extra ArrayList creation was very heavy since this code runs many times with a lot of furnaces on a server (7-10K in my tests). Making this change removed 40ms of tick by itself.

Reply:
Then apply that change to the proper patch we have.